### DISCUSSION TITLE
How to create external modules?

### DISCUSSION THREAD

**Discussion by GerardKetuma** [projen/projen#2545] [Q&A]
Looking for some direction in creating an external module. Specifically, I've been using Snowpack.dev recently and will like to create an external module. It will inherit from the TypeScript module to start.
Are there docs that show what interfaces the module needs to support, or where should I start looking in the projen repo?

Thanks.

**Comment by pgarbe** [comment]
Here are some of my learnings: 
* Start with projen and a `JsiiProject`
* The docs of your project class needs to have @pjid and your project name
* If you inherit from another project and overwrite defaults it's not visible to your users (can be confusing)
* You can use `Component` to structure your project (and re-use things in different projects)
* Local testing works with a local path 

I don't have a public example, but @hoegertn has one: https://github.com/taimos/taimos-projen/

**Comment by GerardKetuma** [comment]
> Here are some of my learnings:
> 
> * Start with projen and a `JsiiProject`
> * The docs of your project class needs to have @pjid and your project name
> * If you inherit from another project and overwrite defaults it's not visible to your users (can be confusing)
> * You can use `Component` to structure your project (and re-use things in different projects)
> * Local testing works with a local path
> 
> I don't have a public example, but @hoegertn has one: https://github.com/taimos/taimos-projen/

Thanks @pgarbe that is exactly what I need to get started. I'll document the process here for someone else to use when I'm done.

**Comment by pgollucci** [comment]
Make a PR that adds to `docs/external-project.md`

On Fri, Jan 8, 2021 at 12:31 PM Gerard Ketuma <notifications@github.com>
wrote:

> Here are some of my learnings:
>
>    - Start with projen and a JsiiProject
>    - The docs of your project class needs to have @pjid and your project
>    name
>    - If you inherit from another project and overwrite defaults it's not
>    visible to your users (can be confusing)
>    - You can use Component to structure your project (and re-use things
>    in different projects)
>    - Local testing works with a local path
>
> I don't have a public example, but @hoegertn <https://github.com/hoegertn>
> has one: https://github.com/taimos/taimos-projen/
>
> Thanks @pgarbe <https://github.com/pgarbe> that is exactly what I need to
> get started. I'll document the process here for someone else to use when
> I'm done.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/projen/projen/issues/471#issuecomment-756896178>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAIL53OOI5RA2BU33ZPEXLSY46QPANCNFSM4VZTE65A>
> .
>


-- 
Philip M. Gollucci
IT Executive and Engineering Leader
http://www.linkedin.com/in/pgollucci/
301.818.0719

**Comment by eladb** [comment]
+1 on creating a doc/

**Comment by GerardKetuma** [comment]
> Make a PR that adds to `docs/external-project.md`

@pgollucci Sounds like a good idea. I'll do a PR to the docs.

**Comment by pgollucci** [comment]
Any update, I may be able to pick up the doc over the next 10 days or so

**Comment by vito-giarrusso** [comment]
This would be fantastic...
I would give it a try, except i'm still struggling with the whole process of creating an external module...
@eladb (@mentioning a maintainer as per the bots comment)

**Comment by pgollucci** [comment]
Well I might be able to take. a crack at it. I did write a module and am
updating now after 12months.  I can't promise an ETA though.

On Thu, Mar 24, 2022 at 9:49 PM Vito Giarrusso ***@***.***>
wrote:

> Would love for this doco to surface...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/projen/projen/issues/471#issuecomment-1078573406>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAIL5352Y5GMAD2FAVUXVLVBULRBANCNFSM4VZTE65A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Philip M. Gollucci
IT Executive and Engineering Leader
http://www.linkedin.com/in/pgollucci/
301.818.0719
